addons/l10n_in_ewaybill_stock/tools/ewaybill_api.py

```diff
@@ -0,0 +1,207 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+import logging
+import contextlib
+
+from datetime import timedelta
+from markupsafe import Markup
+
+from odoo import fields, _
+from odoo.addons.iap import jsonrpc
+from odoo.exceptions import AccessError
+from odoo.addons.l10n_in_edi.models.account_edi_format import DEFAULT_IAP_ENDPOINT, DEFAULT_IAP_TEST_ENDPOINT
+from odoo.addons.l10n_in_edi_ewaybill.models.error_codes import ERROR_CODES
+
+
+_logger = logging.getLogger(__name__)
+
+
+class EWayBillError(Exception):
+
+    def __init__(self, response):
+        self.error_json = self._set_missing_error_message(response)
+        self.error_json.setdefault('odoo_warning', [])
+        self.error_codes = self.get_error_codes()
+        super().__init__(response)
+
+    def _set_missing_error_message(self, response):
+        for error in response.get('error', []):
+            if error.get('code') and not error.get('message'):
+                error['message'] = self._find_missing_error_message(error.get('code'))
+        return response
+
+    @staticmethod
+    def _find_missing_error_message(code):
+        error_message = ERROR_CODES.get(code)
+        return error_message or _("We don't know the error message for this error code. Please contact support.")
+
+    def get_all_error_message(self):
+        return Markup("<br/>").join(
+            [Markup("[%s] %s") % (e.get("code"), e.get("message")) for e in self.error_json.get('error')]
+        )
+
+    def get_error_codes(self):
+        return [e.get("code") for e in self.error_json['error']]
+
+
+class EWayBillApi:
+
+    DEFAULT_HELP_MESSAGE = _(
+        "Somehow this E-waybill has been %s in the government portal before. "
+        "You can verify by checking the details into the government "
+        "(https://ewaybillgst.gov.in/Others/EBPrintnew.asp)"
+    )
+
+    def __init__(self, company):
+        company.ensure_one()
+        self.company = company
+        self.env = self.company.env
+
+    def _ewaybill_connect_to_server(self, url_path, params):
```

Comment:
There is no way to reuse the methods from l10n_in_edi_ewaybill?

Reply:
We can do a refactor in master? `l10n_in_edi_ewaybill` can use this same class

This will be main class for making request in master for both modules (or may be merging it to one) as of I know as per discussion with @jva-odoo
and we do not plan to touch `l10n_in_edi_ewaybill` in stable versions due to productions

Reply:
It is just that the methods in account_edi_format in l10n_in_edi_ewaybill coud be called the way they are.  (company, payload as params).

Reply:
But this class is way better than what we use in `l10n_in_edi_ewaybill` and better error handling (less repetitive) inaccessible with `env` (secured),  I can move this to `l10n_in_edi_ewaybill` if you and @jva-odoo agrees with that I can make changes to the existing `l10n_in_edi_ewaybill`

This is what we are planning for the future (in master), We were going to remove the requests method from `l10n_in_edi_ewaybill` in master and replace it with this class